claim_validate.go

```diff
@@ -0,0 +1,113 @@
+package jwt
+
+import (
+	"errors"
+	"sort"
+	"time"
+)
+
+type JwtPayloadClaimValidator func(tokenClaims *Claims) error
+
+func IssuerValidator(expectedIss string) JwtPayloadClaimValidator {
+	return func(tokenClaims *Claims) error {
+		tokenIss, ok := tokenClaims.String("iss")
+		if !ok {
+			return errors.New("jwt issuer is missing and is required")
+		}
+
+		if tokenIss != expectedIss {
+			return errors.New("invalid issuer claim")
+		}
+
+		return nil
+	}
+}
+
+func SubjectValidator(expectedSub string) JwtPayloadClaimValidator {
+	return func(tokenClaims *Claims) error {
+		tokenSub, ok := tokenClaims.String("sub")
+		if !ok {
+			return errors.New("missing subject claim")
+		}
+
+		if tokenSub != expectedSub {
+			return errors.New("invalid subject claim")
+		}
+
+		return nil
+	}
+}
+
+func AudiencesValidator(expectedAud []string) JwtPayloadClaimValidator {
+	return func(tokenClaims *Claims) error {
+		tokenAud := tokenClaims.Audiences
+		if len(tokenAud) == 0 {
+			return errors.New("missing audience claim")
+		}
+
+		if len(tokenAud) != len(expectedAud) {
+			return errors.New("invalid audience claim")
+		}
+
+		sort.Strings(tokenAud)
+		sort.Strings(expectedAud)
+
+		for i := range tokenAud {
+			if tokenAud[i] != expectedAud[i] {
+				return errors.New("invalid audience claim")
+			}
+		}
+
+		return nil
+	}
+}
+
+func TimeFieldValidator(expectedTime time.Time) JwtPayloadClaimValidator {
+	return func(tokenClaims *Claims) error {
+		if ok := tokenClaims.Valid(expectedTime); !ok {
+			return errors.New("token has expired")
```

Comment:
It could be that the not-before was in the future, which is something else than expiry.